uber/badge_funcs.py

```diff
@@ -98,3 +98,71 @@ def check_unassigned():
+def badge_consistency_check(session):
+    errors = []
+
+    # check 1, see if anything is out of range, or has a duplicate badge number
+    badge_nums_seen = []
+    for attendee in session.query(Attendee)\
+            .filter(Attendee.first_name != '', Attendee.badge_num != 0)\
+            .order_by(Attendee.badge_num).all():
+        out_of_range_error = check_range(attendee.badge_num, attendee.badge_type)
+        if out_of_range_error:
+            msg = '{a.full_name}: badge #{a.badge_num}: {err}'.format(a=attendee, err=out_of_range_error)
+            errors.append(msg)
+
+        if attendee.badge_num in badge_nums_seen:
+            msg = '{a.full_name}: badge #{a.badge_num}: Has been assigned the same badge number of another badge, which is not supposed to happen'.format(a=attendee)
+            errors.append(msg)
+
+        badge_nums_seen.append(attendee.badge_num)
+
+    # check 2: see if there are any gaps in each of the badge ranges
+    for badge_type_val, badge_type_desc in c.BADGE_OPTS:
+        prev_badge_num = -1
+        prev_attendee_name = ""
+
+        for attendee in session.query(Attendee)\
+                .filter_by(badge_type=badge_type_val)\
+                .filter(Attendee.first_name != '', Attendee.badge_num != 0)\
+                .order_by(Attendee.badge_num):
+            if prev_badge_num == -1:
+                prev_badge_num = attendee.badge_num
+                prev_attendee_name = attendee.full_name
+                continue
+
+            if attendee.badge_num - 1 != prev_badge_num:
+                msg = "gap in badge sequence between " + badge_type_desc + " " + \
+                      "badge# " + str(prev_badge_num) + "(" + prev_attendee_name + ")" + " and " + \
+                      "badge# " + str(attendee.badge_num) + "(" + attendee.full_name + ")"
+
+                errors.append(msg)
+
+            prev_badge_num = attendee.badge_num
+            prev_attendee_name = attendee.full_name
+
+    return errors
+
+
+# EXTREME CAUTION. POWERFUL DARK MAGIC. NUCLEAR OPTION. YOU HAVE BEEN WARNED
+# DO NOT USE THIS UNLESS YOU KNOW EXACTLY WHAT YOU ARE DOING!!! i.e. ask Dom, Eli, or Vicki
+# this will re-assign every single badge number for every single attendee to be in the correct range,
+# be EXTREMELY careful with it. it's intended to fix up the database in situations where badge numbers
+# get seriously fubar'd.
+def fixup_all_badge_numbers(session):
+    with c.BADGE_LOCK:
+        for badge_type, badge_range in c.BADGE_RANGES.items():
+            next_available_badge_number = badge_range[0]
+            for attendee in session.query(Attendee)\
+                    .filter_by(badge_type=badge_type)\
+                    .filter(Attendee.first_name != '', Attendee.badge_num != 0)\
+                    .order_by(Attendee.badge_num).all():
+                attendee.badge_num = next_available_badge_number
+                next_available_badge_number += 1
+
+        session.commit()
```

Comment:
Both this function and the previous check function are missing a failure case that we're actually seeing in production right now, which is that some of our staff badges (which start at 1) all got shifted downwards a bit.  So instead of e.g. Brendan / Dom / Paul being 1 / 2 / 3, we ended up with them being like -1 / 0 / 1.  Except that this has kept happening so I think the negative numbers are even lower than that now.

This means that checking for ``Attendee.badge_num != 0`` will actually miss someone whose badge number is 0 but shouldn't be.

I've outlined a long term fix to avoid this problem at https://github.com/magfest/ubersystem/issues/1637

In the meantime, a better approach for this particular function is probably to do something like

```python
for badge_type, badge_range in c.BADGE_RANGES.items():
    if badge_type in c.PREASSIGNED_BADGE_TYPES:
        ...
```

and then instead of ``Attendee.badge_type != 0`` we can say ``Attendee.paid != c.NOT_PAID`` since everyone in every other "paid" state should have a badge number assigned.